From 9ae30d54a158d94fc6e325a62a09a6a7362fc6ed Mon Sep 17 00:00:00 2001
From: Rewryte <129854247+Rewryte@users.noreply.github.com>
Date: Fri, 15 Sep 2023 03:07:13 +0800
Subject: [PATCH 1/2] add visibility update

---
 src/game.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/game.cpp b/src/game.cpp
index 7916b24d0b63e..dfb9a33bd0c04 100644
--- a/src/game.cpp
+++ b/src/game.cpp
@@ -10534,6 +10534,10 @@ bool game::walk_move( const tripoint &dest_loc, const bool via_ramp, const bool
     if( moving ) {
         cata_event_dispatch::avatar_moves( old_abs_pos, u, m );
 
+        // Update map and visibility caches
+        m.build_map_cache( u.posz() );
+        m.update_visibility_cache( u.posz() );
+
         // Add trail animation when sprinting
         if( get_option<bool>( "ANIMATIONS" ) && u.is_running() ) {
             if( u.posy() < oldpos.y ) {

From 95618086b24945b746d40e215fe37d52bfd1a3ef Mon Sep 17 00:00:00 2001
From: Rewryte <129854247+Rewryte@users.noreply.github.com>
Date: Fri, 15 Sep 2023 03:42:46 +0800
Subject: [PATCH 2/2] vis update when peeking

---
 src/game.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/game.cpp b/src/game.cpp
index dfb9a33bd0c04..28388a685294d 100644
--- a/src/game.cpp
+++ b/src/game.cpp
@@ -6276,6 +6276,10 @@ void game::peek( const tripoint &p )
     const bool is_standup_peek = is_same_pos && u.is_crouching();
     tripoint center = p;
 
+    // Update map and visibility caches
+    m.build_map_cache( p.z );
+    m.update_visibility_cache( p.z );
+
     look_around_result result;
     const look_around_params looka_params = { true, center, center, false, false, true };
     if( is_standup_peek ) {   // Non moving peek from crouch is a standup peek
